add: `sqlmesh` metrics

Jabolol opened the pull request:
WIP

ravenac95 commented on warehouse/metrics_mesh/oso_metrics/active_developers.sql:
```diff
@@ -0,0 +1,14 @@
+  COUNT(distinct events.from_artifact_id) as amount
+from metrics.events_daily_to_artifact as events
+where events.event_type = 'COMMIT_CODE'
+  and events.bucket_day BETWEEN @metrics_start(DATE) AND @metrics_end(DATE)
```
```suggestion
  and events.bucket_day BETWEEN @metrics_start('DATE') AND @metrics_end('DATE')
```

Turns out I made an error here with my initial docs of this. It _should_ be a string input here. Eventually this will be used as a way to figure out the type casting and the macro needs this value as a string.

ravenac95 commented on warehouse/metrics_mesh/oso_metrics/developer_states.sql:
```diff
@@ -0,0 +1,50 @@
+  current.amount as current_contributions,
+  coalesce(prev.amount, 0) as previous_contributions
+from
+  @metrics_peer_ref(
```
This doesn't currently work but I _love_ this idea. I think it actually makes a lot of sense but in order to accomplish this right now we'd actually need a CTE for the `current` and the `prev`. Much like I do in here: 

* `latest` https://github.com/opensource-observer/oso/blob/main/warehouse/metrics_mesh/oso_metrics/change_in_developers.sql#L1
* `prev` - https://github.com/opensource-observer/oso/blob/main/warehouse/metrics_mesh/oso_metrics/change_in_developers.sql#L22

So you will need a couple things here: 

* You need to use the `@relative_window_sample_date` macro (I've just documented this yesterday I hope it make sense)
* And you likely need the `@metrics_entity_type_col` macro as well, which has been documented. However, the magic here is hopefully not _too_magical.

Jabolol replied:
Will take into account!

ravenac95 commented on warehouse/metrics_mesh/oso_metrics/developer_states.sql:
```diff
@@ -0,0 +1,50 @@
+select
+  @metrics_sample_date(current.metrics_sample_date) as metrics_sample_date,
+  current.event_source,
+  @metrics_entity_type_col(
```
:+1: correct use of this macro :)

Jabolol commented:
Implemented all metrics on #2420 except `new_contributor_count_6_months` and `fulltime_developer_average_6_months`, that need an extra event table. Removed `Libin` model for the moment, needs more thought and another event table as well.

ravenac95 commented:
@Jabolol FYI i'm going to take over this PR! I'm going to remove some of the things related to first time/last time as those aren't going to fit into these timeseries metrics. On a subsequent PR I'll make the additional model for new contributor events and such.

ravenac95 commented on warehouse/metrics_mesh/oso_metrics/first_commit_date.sql:
```diff
@@ -0,0 +1,15 @@
+select @metrics_sample_date(events.bucket_day) as metrics_sample_date,
```
Just as a note, first commit date would only be useful as an overall KPI for any given user + repo combination. So this would be some kind of intermediate table that would contain this. All of the timeseries metrics are incremental so to get first commit date we'd still need to look at all of time to get the correct "first" commit date.

ravenac95 replied:
Ah I realize now, we did discuss this but I think my thinking here has changed. If we do this as a rolling set of metrics, I don't think doing this as a rolling set of metrics will improve performance here. I believe it will still be quite an expensive call regardless seeing as you will be processing all history. I might be wrong but let's try the simplest thing first for "first/last" commits which is to compute them as point in time metrics. I am thinking through solving churn but will also implement new contributors elsewhere.

ravenac95 commented on warehouse/metrics_mesh/oso_metrics/repositories.sql:
```diff
@@ -0,0 +1,25 @@
+select @metrics_sample_date(events.bucket_day) as metrics_sample_date,
```
Hrm. shoot I'm realizing this is likely on me and I didn't notice this earlier. The point-in-time metrics like repository counts and star counts actually need to be models _outside_ of time series metrics. I'll implement those now.

ryscheng replied:
@ravenac95 fwiw, we probably want both the all-time counts as well as rolling-window versions (e.g. star_count_6_months)

ravenac95 replied:
oh cool ok i'll keep these in here then

ravenac95 replied:
@ryscheng for the point in time metrics for repos, the most accurate measure is actually the ossd_repositories dagster asset. So I created a dbt model to represent point in time metrics of those kinds.

ryscheng replied:
ya for these things we're still using code_metrics_by_project in dbt FWIW
